### DISCUSSION TITLE
Unable to readFromDevice, but  onDeviceRead and write works

### DISCUSSION THREAD

**Discussion by DamianoBarone** [kenjdavidson/react-native-bluetooth-classic#168] [General]
Hi,
I have a quick question that probalby is a easy fix.
I was able to send a discover, connect and write with RNBluetoothClassic. However, I'm stuck with the read. I'm quite new to react native, so maybe it's just that i don't know how to print the return String (hopefully)
I will put the more intestersting code:

```
async function connectDev() {
  try {
    console.log("connectDev");
    await RNBluetoothClassic.connectToDevice(deviceAddr,
      {
        CONNECTOR_TYPE: "rfcomm",
        DELIMITER: "\n",
        DEVICE_CHARSET: "ascii",
      });
    await RNBluetoothClassic.writeToDevice(deviceAddr, "Hello\n");
  } catch (error) {
    console.log('Error:', error);

  }
}
function readMessage(){
  console.log('readMessage');
  var data = RNBluetoothClassic.availableFromDevice(deviceAddr);
  console.log(data);

  var value = RNBluetoothClassic.readFromDevice(deviceAddr);
  console.log(value);
  console.log(JSON.stringify(value, null, 2));
}
function mainApp(props) {
  useEffect(() => {
  this.subscription = RNBluetoothClassic.onDeviceRead(deviceAddr, readMessage());
...
```

It doesn't work readFromDevice and availableFromDevice but readMessage is called. If i print directly the variable, it says is an object, if i try with stringify it prints:

```
 LOG  {
  "_U": 0,
  "_V": 0,
  "_W": null,
  "_X": null
}
```

I will really appreciate any suggestion :)
I couldn't find any example how to use that functions. 
Note:anyway I really appreciate your work !

**Comment by kenjdavidson** [comment]
Almost all the functions return promises, meaning you'll need to `await` even the read.  

```
var data = await RNBluetoothClassic.availableFromDevice(deviceAddr);
```

ps.  Documentation

https://kenjdavidson.com/react-native-bluetooth-classic/react-native/rn-bluetooth-device/#read

**Comment by kenjdavidson** [comment]
A better method, which I've been attempting to push is:

```
try {
   let device = await RNBluetoothClassic.connectToDevice(address);  // Returns BlutoothDevice
   let data = await device.read();
   device.write(data);
} catch () {}
```

**Comment by DamianoBarone** [comment]
Hi,
thank you for the quick answer!
I have tried first
```
async function readMessage(){
  console.log('readMessage');
  var available = await RNBluetoothClassic.availableFromDevice(deviceAddr);
  console.log(available);

  var value = await RNBluetoothClassic.readFromDevice(deviceAddr);

  console.log(JSON.stringify(value, null, 2));
}
```

the resul was not good:
```
 LOG  0
 LOG  null
```

I have removed and add a async function
```
async function discover2() {
  try {
    let device = await RNBluetoothClassic.connectToDevice("F0:08:D1:C9:C7:3E");  // Returns BlutoothDevice
    if (device.isConnected)
    {
      console.log("end connectDev");
      device.write("hello\n");
  
      let data = await device.read();
      console.log (data);
    }

 } catch (error) {
    console.log('Error:', error);
  }
}
```

```
 LOG  isBluetoothEnabled true
 LOG  isBluetoothAvailable true
 LOG  end connectDev
 LOG  null
```

So the issue is still there...
My other device is just a ESP32 that when receive a message, it sends automatically a json. In the ESP32 I can see the write working properly. It doesn't make sense that the read is not working.... I send this `{'sensor':'HEART SENSOR','data':{'bpm':71,'live': 0}}`

any other suggestion?

**Comment by kenjdavidson** [comment]
Looking at your information, I can see the following:

```
var available = await RNBluetoothClassic.availableFromDevice(deviceAddr);
// LOG  0

var value = await RNBluetoothClassic.readFromDevice(deviceAddr);
// LOG null
```

Available says there is nothing there, so you're getting `null` when you receive data.  This seems to be working as intended, 0 data = nothing returned.

I would assume that your ESP code is not deliminating the messages with a `\n` for example:

```
{'sensor':'HEART SENSOR','data':{'bpm':71,'live': 0}}\n
```

needs to be sent in order for it to be processed as a message.

**Reply by priya-kato** [reply]
Yes actually i got this perfect one.

**Reply by kenjdavidson** [reply]
I still think you're combining `onRead` and manual read and this is the problem.  You should NEVER have:
- `onDeviceRead`
- `readFromDevice` 

in the same application.  These two don't work well together, so if you're using both, you're not doing it right.

**Comment by DamianoBarone** [comment]
Hi,
but that function readMessage, it's a callabck from onDeviceRead, that it should happen when we have data to read it. Or am I wrong?
I solved to put a delay on the read (i'm waiting 5 sec, probably i can reduce it).
thank you very much! 
you deserve a gold statue.

**Comment by kenjdavidson** [comment]
I have no idea what you're doing (sorry).

The code that you've posted is using the MANUAL read logic:
- get available
- readFromDevice

which should NOT be coupled with the automatic read logic:
- device = connectToDevice();
- device.onRead((data) -> handleRead(data));

There is nothing in the code you've posted that leads me to believe you're using the on read listener.

**Comment by kenjdavidson** [comment]
If you're doing a:

```
function readData() {
   getAvailable();
   readFromDevice();
}
```

then somewhere else

```
device.onread((data) -> readData());
```

Then you're:
- being sent the data 
- Ignoring it
- attempting to read empty data (since you already received it)

You don't want to do this, using manual read with automatic read will do some weird stuff.

**Comment by DamianoBarone** [comment]
Ah no worries at all.
if you see my first comment 
`function mainApp(props) {
  useEffect(() => {
  this.subscription = RNBluetoothClassic.onDeviceRead(deviceAddr, readMessage());`


and inside readMessage i'm doing the manual reading.

 if want use the onDeviceRead  as a callback  to get the data how should i do it?
I would like have  just a function that is called everytime I have data to read.
i hope make sense

**Reply by kenjdavidson** [reply]
> I would like have just a function that is called everytime I have data to read.

This doesn't exist.

**Comment by kenjdavidson** [comment]
Ya, that's not right.

I would suggest following the examples on:
- the documentation site
- the sample application

They are fully working, they are something like this:

```
device.onRead((data) -> console.log(data));
```

which should get you started.

**Reply by DamianoBarone** [reply]
anyway thank you!
again awesome work!

**Reply by nickanimate** [reply]
Did the issue get resolved?

**Reply by DamianoBarone** [reply]
Yes, I’m doing active reading with a timer.
Thank you very much

**Reply by nickanimate** [reply]
That's great. Can you give me the sample how you are doing it? I am getting null value when reading, no matter what I do

**Reply by DamianoBarone** [reply]
Just do the example in the documentation or in the comment 
https://github.com/kenjdavidson/react-native-bluetooth-classic/discussions/168#discussioncomment-2175808

**Reply by kenjdavidson** [reply]
From my understanding of this, the problem is combining manual (read from device) and event (on Read device) which will not work.

On read event will get the data and ignore it. 
Read from device will return null, as the actual data was sent to the event.

Not sure if the docs or samples are confusing, but there is no event to notify of a manual read. Although I'm starting to see the use case for this.

So maybe

1) On Device Read would send data

2) On device data would just notify of new data
Read from device would manual read.

Thoughts?

**Comment by Giacomo90** [comment]
I have the same problem, I can connect, I can write, but I don't receive anything.
I try 
device.onRead((rx)=>{console.log(rx)}) not work.
device.read() receive only null.

**Reply by kztneda** [reply]
have you solved this? I have the same issue

**Reply by kenjdavidson** [reply]
The majority of the time this is a delimiter issue.  The default is `\n\r` or something, and you're not sending that.  Without the delimiter the message buffer is never chunked to individual messages and it will always return `null`.  You can debug this on the Native side.  I would suggest confirming:
- What your connection options are `device.connect({DELIMITER: '\n'})` if you want to specify it
- What is the delimiter you're actually sending?  Is it `\r` or `\n` or `\n\r` 
- Make sure they match.
